### PR TITLE
CO-2269 Dissocier la communication Welcome Muskathlon d'un nouveau Portal User

### PR DESCRIPTION
**Pull request by jo-bloiz-ornot**
Wizard that will let you choose if you want to create a communication and let you choose which template (for now there is only Muskathlon template)

### CODE REVIEW

File: partner_communication_switzerland/wizards/portal_wizard.py [hunk 1]
```diff
@@ -9,25 +9,60 @@
...
-                'object_ids': user.id,
-                'config_id': config.id
-            })
+        res = super(PortalWizardUser, self).action_apply()
+
+        # to not create a communication
+        # you should give context
+        # { 'create_uid_communication': False }
+        # else it will create a communication
+        if self.env.context['create_communication_with_uid'] is None:
```
Comment (ecino) [inline partner_communication_switzerland/wizards/portal_wizard.py]: Instead of reading this value from the context, it's better and more clear to directly add a field in the wizard 'portal.wizard.user' model. `invitation_config_id = fields.Many2one('partner.communication.config')`. With that you can here directly test `if self.invitation_communication_id:`

File: partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+          'the login link to the user ?'), default=True)
+
+    config_id = fields.Many2one(
+        'partner.communication.config', _('choose a communication config'),
+        domain="[('model', '=', 'res.users')]"
+    )
+
+    @api.multi
+    def button_create_portal_user(self):
+        portal = self.env['portal.wizard'].create({})
```
Comment (ecino) [inline partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py]: If you add the field `invitation_config_id` in the 'portal.wizard.user' model, you can here pass it in the creation values `{'invitation_config_id': self.config_id}`

File: partner_communication_switzerland/wizards/portal_wizard.py [hunk 3]
```diff
@@ -9,25 +9,60 @@
...
+
         return res
+
+    @api.multi
+    def _create_uid_communication(self):
+        if not self.env.user.email:
+            raise UserError(_('You must have an email address in'
+                              ' your User Preferences to send emails.'))
+
+        config_id_id = self.env.context['config_id_id'] or \
```
Comment (ecino) [inline partner_communication_switzerland/wizards/portal_wizard.py]: If you add the field, here the communication would be simply `self.invitation_config_id`. I don't think that you should catch a case where the user wants to send something but doesn't select any communication rule. It makes no sense, either they choose one to send or we don't send nothing.

File: partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+
+class ResPartnerCreatePortalWizard(models.TransientModel):
+    """ creation of a portal user wizard will send a email
+        with the identifier if the user used the checkbox """
+    _name = 'res.partner.create.portal.wizard'
+    _description = 'take a partner and make it a odoo user'
+
+    send_communication = fields.Boolean(
+        _('do you want to send a mail with '
```
Comment (ecino) [inline partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py]: Field strings don't need to be translated with the python function, they are automatically included in translations. Try to keep them short or it will become too long in the user interface. Here I propose simply `send_communication = fields.Boolean('Send an e-mail invitation')`

File: partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+          'the login link to the user ?'), default=True)
+
+    config_id = fields.Many2one(
+        'partner.communication.config', _('choose a communication config'),
```
Comment (ecino) [inline partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py]: 'Choose a communication'

File: partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        for user in no_mail:
+            partner = user.partner_id
+            user.email = partner.firstname[0].lower() + \
+                partner.lastname.lower() + '@cs.local'
+
+        # define if needed to create a communication
+        bool_communication_with_uid = \
+            (self.send_communication and self.config_id)
+
+        ctx = {
```
Comment (ecino) [inline partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py]: No need to put a custom context if you pass the config via the field

File: partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+    config_id = fields.Many2one(
+        'partner.communication.config', _('choose a communication config'),
+        domain="[('model', '=', 'res.users')]"
+    )
+
+    @api.multi
+    def button_create_portal_user(self):
+        portal = self.env['portal.wizard'].create({})
+        portal.onchange_portal_id()
```
Comment (ecino) [inline partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py:37]: Add a comment here to explain that the following lines are for allowing registering users even if they don't have an e-mail address.

File: partner_communication_switzerland/wizards/portal_wizard.py [hunk 8]
```diff
@@ -9,25 +9,60 @@
...
+            user_id = wizard_line.user_id
+            partner_id = user_id.partner_id
+
+            self.env['partner.communication.job'].create({
+                'partner_id': partner_id.id,
+                'object_ids': user_id.id,
+                'config_id': config_id_id
+            })
+
+        return True
```
Comment (ecino) [inline partner_communication_switzerland/wizards/portal_wizard.py]: Instead of returning True, you can return the communication created in order to use it directly. `return self.env['partner.communication.job'].create`

File: partner_communication_switzerland/wizards/portal_wizard.py [hunk 9]
```diff
@@ -9,25 +9,60 @@
...
+        # { 'create_uid_communication': False }
+        # else it will create a communication
+        if self.env.context['create_communication_with_uid'] is None:
+            bool_communication_with_uid = True
+        else:
+            bool_communication_with_uid = \
+                self.env.context['create_communication_with_uid']
+
+        if bool_communication_with_uid:
+            self._create_uid_communication()
```
Comment (ecino) [inline partner_communication_switzerland/wizards/portal_wizard.py]: Here you can add inside res the communication created (or just the id maybe if it doesn't work). `res['communication'] = self._create_uid_communication()`. `_create_uid_communication` has of course to return the communication record.

File: partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+            'create_communication_with_uid': bool_communication_with_uid,
+            'config_id_id': self.config_id.id
+        }
+        portal.with_context(ctx).action_apply()
+        no_mail.mapped('partner_id').write({'email': False})
+
+        action = None
+        if bool_communication_with_uid:
+            # get same communication create in portal.action_apply()
+            same_job_search = [
```
Comment (ecino) [inline partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py]: If your function `action_apply` puts the communication inside the result, here you can simply retrieve it by doing `communication = action_result['communication']`

File: partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+
+        # define if needed to create a communication
+        bool_communication_with_uid = \
+            (self.send_communication and self.config_id)
+
+        ctx = {
+            'create_communication_with_uid': bool_communication_with_uid,
+            'config_id_id': self.config_id.id
+        }
+        portal.with_context(ctx).action_apply()
```
Comment (ecino) [inline partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py]: assign the result in order to retrieve the generated communication : `action_result = portal.action_apply()`

**Review by ecino — CHANGES_REQUESTED** [review]
Very good changes, thank you! Here are some suggestions in order to improve how parameters and results are passed between the different objects and methods.

File: partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py (hunk 2 above)
Comment (jo-bloiz-ornot) [reply partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py]: The problem is that it's not a portal wizard user but just a portal wizard, I think overriding the method (onchange_id) that create portal user is possible but it's will lose the adaptability of the oca module

Comment (jo-bloiz-ornot) [reply partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py]: I think I can do it without losing changes of oca module, I tell you later

File: partner_communication_switzerland/wizards/portal_wizard.py [hunk 12]
```diff
@@ -9,25 +9,73 @@
...
-        config = self.env.ref(
-            'partner_communication_switzerland.portal_welcome_config')
-        for user in self.mapped('user_ids.user_id'):
-            comm_obj.create({
-                'partner_id': user.partner_id.id,
-                'object_ids': user.id,
-                'config_id': config.id
-            })
-        return res
+        self.ensure_one()
```
Comment (ecino) [inline partner_communication_switzerland/wizards/portal_wizard.py]: You removed the call to the super which here can remove some useful features. Can you add it back?

**Review by ecino — CHANGES_REQUESTED** [review]

File: partner_communication_switzerland/wizards/portal_wizard.py (hunk 12 above)
Comment (jo-bloiz-ornot) [reply partner_communication_switzerland/wizards/portal_wizard.py]: No, because the parent didn't return the result of the action_apply method

```
    @api.multi
    def action_apply(self):
        self.ensure_one()
        self.user_ids.action_apply()
        return {'type': 'ir.actions.act_window_close'}
```
Override had to be done there

Comment (ecino) [reply partner_communication_switzerland/wizards/portal_wizard.py]: The problem is that this code will never be executed :
https://github.com/compassionch/compassion-modules/blob/bac7e809eb6deabea16c080698f84122601a69d9/crm_compassion/wizards/portal_wizard.py#L21
You should call the super even if it doesn't return what we need and then return what you need after. Maybe here some conception needs to be changed to ensure we don't lose the code of crm_compassion.

Comment (jo-bloiz-ornot) [reply partner_communication_switzerland/wizards/portal_wizard.py]: ok

Just one question, is the fact that the method portal.wizard.user action_apply() will be call once in the super method then again just to get the result is a problem ?

Comment (jo-bloiz-ornot) [reply partner_communication_switzerland/wizards/portal_wizard.py]: I will try with a additional field in portal.wizard.user to not call twice the method

Comment (jo-bloiz-ornot) [reply partner_communication_switzerland/wizards/portal_wizard.py]: it's a success

File: partner_communication_switzerland/wizards/portal_wizard.py [hunk 13]
```diff
@@ -9,25 +9,65 @@
...
-                'object_ids': user.id,
-                'config_id': config.id
-            })
+        res = super(PortalWizardUser, self).action_apply()
+
+        self.mapped('partner_id').signup_prepare()
+
+        # if the context is not given, create a communication
+        # else do as context say
+        if self.env.context['create_communication'] is None or \
```
Comment (ecino) [inline partner_communication_switzerland/wizards/portal_wizard.py]: I don't understand why you create a communication when context is None. To me, it sounds more logic to create the communication only when we are told to do so. This would mean `if self.env.context.get('create_communication'):`

File: partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+            user.email = partner.firstname[0].lower() + \
+                partner.lastname.lower() + '@cs.local'
+
+        portal.with_context({
+            'create_communication': self.create_communication
+        }).action_apply()
+
+        no_mail.mapped('partner_id').write({'email': False})
+
+        action = None
```
Comment (ecino) [inline partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py]: `action = True` is better to return.

File: partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+                'view_type': 'form',
+                'view_mode': 'tree,form',
+                'domain': [('id', 'in', uid_communication.ids)],
+                'flags': {'action_buttons': True}
+            }
+        return action
+
+    @api.multi
+    def button_cancel(self):
+        return None
```
Comment (ecino) [inline partner_communication_switzerland/wizards/res_partner_create_portal_wizard.py]: return True (means ok).

File: partner_communication_switzerland/wizards/portal_wizard.py [hunk 16]
```diff
@@ -9,25 +9,71 @@
...
-            comm_obj.create({
-                'partner_id': user.partner_id.id,
-                'object_ids': user.id,
-                'config_id': config.id
-            })
+        res = super(PortalWizardUser, self).action_apply()
+
+        self.mapped('partner_id').signup_prepare()
+
+        if self.env.context['create_communication']:
```
Comment (ecino) [inline partner_communication_switzerland/wizards/portal_wizard.py]:
```suggestion
        if self.env.context.get('create_communication'):
```
This avoids the code execution to fail in case `create_communication` is not present in the context

File: partner_communication_switzerland/wizards/portal_wizard.py [hunk 17]
```diff
@@ -9,25 +9,71 @@
...
+            for wizard_line in self:
+                wizard_line.create_uid_communication()
+        else:
+            for wizard_line in self:
```
Comment (ecino) [inline partner_communication_switzerland/wizards/portal_wizard.py]: Why did you add this? I'm not sure we want to automatically activate the account if we don't send a communication.

**Review by ecino — CHANGES_REQUESTED** [review]

File: partner_communication_switzerland/wizards/portal_wizard.py (hunk 16 above)
Comment (jo-bloiz-ornot) [reply partner_communication_switzerland/wizards/portal_wizard.py]: good to know, thank you

File: partner_communication_switzerland/wizards/portal_wizard.py (hunk 17 above)
Comment (jo-bloiz-ornot) [reply partner_communication_switzerland/wizards/portal_wizard.py]: I was afraid to give to much work to the user to activate manually the account, because if you dont send a email with a link to activate the account, who gonna activate it ?

Comment (ecino) [reply partner_communication_switzerland/wizards/portal_wizard.py]: In fact we register some users that we never want to give access to the account, only to be able to assign them to some events or other objects (but this is intern stuff and we don't communicate to them). So we don't wan't that everyone is activated automatically. I think you can remove this part.

Comment (jo-bloiz-ornot) [reply partner_communication_switzerland/wizards/portal_wizard.py]: alright, it's gonna be done

**Review by ecino — APPROVED** [review]
